In [1]:
#importing librarys 

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

## Problem 01


In [2]:
file_name = 'tidy_data_table_9_use.csv'

In [3]:
#import file
tub = pd.read_csv( file_name)

In [4]:
tub.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5769 entries, 0 to 5768
Data columns (total 18 columns):
 #   Column   Non-Null Count  Dtype  
---  ------   --------------  -----  
 0   country  5746 non-null   object 
 1   year     5769 non-null   int64  
 2   f_014    2378 non-null   float64
 3   f_1524   2394 non-null   float64
 4   f_2534   2400 non-null   float64
 5   f_3544   2398 non-null   float64
 6   f_4554   2401 non-null   float64
 7   f_5564   2394 non-null   float64
 8   f_65     2396 non-null   float64
 9   f_u      272 non-null    float64
 10  m_014    2381 non-null   float64
 11  m_1524   2407 non-null   float64
 12  m_2534   2408 non-null   float64
 13  m_3544   2415 non-null   float64
 14  m_4554   2421 non-null   float64
 15  m_5564   2414 non-null   float64
 16  m_65     2408 non-null   float64
 17  m_u      273 non-null    float64
dtypes: float64(16), int64(1), object(1)
memory usage: 811.4+ KB


In [5]:
tub.head()

country  year  f_014  f_1524  f_2534  f_3544  f_4554  f_5564  f_65  f_u  \
0      AD  1989    NaN     NaN     NaN     NaN     NaN     NaN   NaN  NaN   
1      AD  1990    NaN     NaN     NaN     NaN     NaN     NaN   NaN  NaN   
2      AD  1991    NaN     NaN     NaN     NaN     NaN     NaN   NaN  NaN   
3      AD  1992    NaN     NaN     NaN     NaN     NaN     NaN   NaN  NaN   
4      AD  1993    NaN     NaN     NaN     NaN     NaN     NaN   NaN  NaN   

   m_014  m_1524  m_2534  m_3544  m_4554  m_5564  m_65  m_u  
0    NaN     NaN     NaN     NaN     NaN     NaN   NaN  NaN  
1    NaN     NaN     NaN     NaN     NaN     NaN   NaN  NaN  
2    NaN     NaN     NaN     NaN     NaN     NaN   NaN  NaN  
3    NaN     NaN     NaN     NaN     NaN     NaN   NaN  NaN  
4    NaN     NaN     NaN     NaN     NaN     NaN   NaN  NaN

In [6]:
# saving column names so I can split on _
col_names = tub.columns.to_list()
del col_names[0:2]

In [7]:
col_names

['f_014',
 'f_1524',
 'f_2534',
 'f_3544',
 'f_4554',
 'f_5564',
 'f_65',
 'f_u',
 'm_014',
 'm_1524',
 'm_2534',
 'm_3544',
 'm_4554',
 'm_5564',
 'm_65',
 'm_u']

In [8]:
# melt data to get if from wide to long
lf = tub.melt( id_vars=['country','year'], value_vars=col_names )

In [9]:
lf.head(20)

country  year variable  value
0       AD  1989    f_014    NaN
1       AD  1990    f_014    NaN
2       AD  1991    f_014    NaN
3       AD  1992    f_014    NaN
4       AD  1993    f_014    NaN
5       AD  1994    f_014    NaN
6       AD  1996    f_014    0.0
7       AD  1997    f_014    0.0
8       AD  1998    f_014    NaN
9       AD  1999    f_014    0.0
10      AD  2000    f_014    NaN
11      AD  2001    f_014    NaN
12      AD  2002    f_014    0.0
13      AD  2003    f_014    0.0
14      AD  2004    f_014    0.0
15      AD  2005    f_014    0.0
16      AD  2006    f_014    0.0
17      AD  2007    f_014    NaN
18      AD  2008    f_014    0.0
19      AE  1980    f_014    NaN

In [10]:
# split variable by '_' and save back into df
lf[['sex', 'age']] = lf.variable.str.split( '_', expand=True )

In [11]:
# drop variable b/c its redundant now
lf.drop(['variable'], axis=1, inplace=True)

In [13]:
lf.shape

(92304, 5)

How many rows and columns does your final "tidy" data set consist of?
5 columns and 92,304 rows

In [14]:
len(pd.unique(lf['country'])) 

213

In [15]:
len(pd.unique(tub['country'])) 

213

How many unique `country` values are there? Is this consistent with the number of unique `country` values in the original data set?
213 unique country values, it is consistent with the original dataset

In [16]:
len(pd.unique(lf['year'])) 

29

In [17]:
len(pd.unique(tub['year'])) 

29

How many unique `year` values are there? How many rows are associated with each year in the final "tidy" data set?
29 unique years, it is consistent with the orginal dataset.

## Problem 03

In [18]:
lf.isna().groupby(['year']).aggregate(total_number = ('value', 'sum'))

total_number
year               
False         58144

Which year has the most missing values? Which year has the fewest missing values?
Could not figure it out... What is above was my best attempt. 

In [19]:
country_sex = lf.groupby(['country','year', 'sex']).\
aggregate(total_number = ('value', 'sum')).\
reset_index()

In [20]:
filter = ['AO', 'AR']
subset = country_sex[country_sex.country.isin(filter)]

In [21]:
subset.sort_values(['sex','total_number'], ascending = (True, False))

country  year sex  total_number
430      AO  2008   f       10681.0
426      AO  2006   f       10542.0
422      AO  2004   f       10360.0
424      AO  2005   f       10245.0
428      AO  2007   f       10141.0
..      ...   ...  ..           ...
455      AR  1991   m           0.0
457      AR  1992   m           0.0
459      AR  1993   m           0.0
461      AR  1994   m           0.0
463      AR  1995   m           0.0

[116 rows x 4 columns]

Which year has the most total cases (across all age groups) for males in the countries with codes AO and AR?
2008

In [22]:
filter = ['AF', 'AM']
subset = country_sex[country_sex.country.isin(filter)]
subset.sort_values(['sex','total_number'], ascending = (True, False))

country  year sex  total_number
134      AF  2007   f        9031.0
136      AF  2008   f        8915.0
132      AF  2006   f        8520.0
130      AF  2005   f        6818.0
128      AF  2004   f        5587.0
..      ...   ...  ..           ...
311      AM  1990   m           0.0
313      AM  1991   m           0.0
315      AM  1992   m           0.0
317      AM  1993   m           0.0
319      AM  1994   m           0.0

[106 rows x 4 columns]

In [23]:
country_sex.to_csv('tb_tidy.csv', header=True, index=False)

Which year has the most total cases (across all age groups) for females in the countries with codes AF and AM?
2008

## Problem 03

In [24]:
file_name_close = 'daily_close_prices.csv'
file_name_high = 'daily_high_prices.csv'
file_name_low = 'daily_low_prices.csv'
file_name_open = 'daily_open_prices.csv'
file_name_key = 'stock_info.csv'

In [25]:
dfclose = pd.read_csv( file_name_close)
dfhigh = pd.read_csv( file_name_high)
dflow = pd.read_csv( file_name_low)
dfopen = pd.read_csv( file_name_open)
dfkey = pd.read_csv( file_name_key)

In [26]:
dfclose.head()

date  AAPL_close   AMZN_close  GME_close   KO_close  TSLA_close  \
0  2020-09-01  134.179993  3499.120117       7.65  49.119999  475.049988   
1  2020-09-02  131.399994  3531.449951       7.71  51.189999  447.369995   
2  2020-09-03  120.879997  3368.000000       7.82  50.480000  407.000000   
3  2020-09-04  120.959999  3294.620117       7.65  51.040001  418.320007   
4  2020-09-08  112.820000  3149.840088       7.70  49.810001  330.209991   

    VZ_close  
0  59.160000  
1  60.529999  
2  60.610001  
3  60.480000  
4  59.959999

In [27]:
# saving column names
col_names = dfclose.columns.to_list()
del col_names[0:2]

In [28]:
col_names

['AMZN_close', 'GME_close', 'KO_close', 'TSLA_close', 'VZ_close']

In [29]:
# melting data
lf_close = dfclose.melt( id_vars=['date'], value_vars=col_names)

In [30]:
lf_close.head()

date    variable        value
0  2020-09-01  AMZN_close  3499.120117
1  2020-09-02  AMZN_close  3531.449951
2  2020-09-03  AMZN_close  3368.000000
3  2020-09-04  AMZN_close  3294.620117
4  2020-09-08  AMZN_close  3149.840088

In [31]:
# splitting variable by '_' an dadding it to dataframe
lf_close[['symbol', 'type']] = lf_close.variable.str.split( '_', expand=True )
lf_close.drop(['variable'], axis=1, inplace=True)

In [32]:
lf_close.head()

date        value symbol   type
0  2020-09-01  3499.120117   AMZN  close
1  2020-09-02  3531.449951   AMZN  close
2  2020-09-03  3368.000000   AMZN  close
3  2020-09-04  3294.620117   AMZN  close
4  2020-09-08  3149.840088   AMZN  close

In [33]:
# doing same to the other datasets
col_names = dfhigh.columns.to_list()
del col_names[0:2]
lf_high = dfhigh.melt( id_vars=['date'], value_vars=col_names)
lf_high[['symbol', 'type']] = lf_high.variable.str.split( '_', expand=True )
lf_high.drop(['variable'], axis=1, inplace=True)

col_names = dflow.columns.to_list()
del col_names[0:2]
lf_low = dflow.melt( id_vars=['date'], value_vars=col_names)
lf_low[['symbol', 'type']] = lf_low.variable.str.split( '_', expand=True )
lf_low.drop(['variable'], axis=1, inplace=True)

col_names = dfopen.columns.to_list()
del col_names[0:2]
lf_open = dfopen.melt( id_vars=['date'], value_vars=col_names)
lf_open[['symbol', 'type']] = lf_open.variable.str.split( '_', expand=True )
lf_open.drop(['variable'], axis=1, inplace=True)

In [34]:
#variables = [dfclose, dfhigh, dflow, dfopen]


#for i in variables:
 #   col_names = i.columns.to_list()
  #  del col_names[0:2]
   # i = i.melt( id_vars=['date'], value_vars=col_names)
    #i[['stock', 'type']] = i.variable.str.split( '_', expand=True )
    #i.drop(['variable'], axis=1, inplace=True)


In [35]:
lf_low.head()

date        value symbol type
0  2020-09-01  3467.000000   AMZN  low
1  2020-09-02  3486.689941   AMZN  low
2  2020-09-03  3303.000000   AMZN  low
3  2020-09-04  3111.129883   AMZN  low
4  2020-09-08  3130.000000   AMZN  low

In [36]:
lf_open.head()

date        value symbol  type
0  2020-09-01  3489.580078   AMZN  open
1  2020-09-02  3547.000000   AMZN  open
2  2020-09-03  3485.000000   AMZN  open
3  2020-09-04  3318.000000   AMZN  open
4  2020-09-08  3144.000000   AMZN  open

In [37]:
# appending all dataframes together - the key
df = lf_close.append([lf_high, lf_open])

In [38]:
df.head(10)

date        value symbol   type
0  2020-09-01  3499.120117   AMZN  close
1  2020-09-02  3531.449951   AMZN  close
2  2020-09-03  3368.000000   AMZN  close
3  2020-09-04  3294.620117   AMZN  close
4  2020-09-08  3149.840088   AMZN  close
5  2020-09-09  3268.610107   AMZN  close
6  2020-09-10  3175.110107   AMZN  close
7  2020-09-11  3116.219971   AMZN  close
8  2020-09-14  3102.969971   AMZN  close
9  2020-09-15  3156.129883   AMZN  close

In [39]:
df = df.pivot_table(index=['symbol', 'date'], columns='type', values='value').reset_index()


In [40]:
df.head()

type symbol        date        close         high         open
0      AMZN  2020-09-01  3499.120117  3513.870117  3489.580078
1      AMZN  2020-09-02  3531.449951  3552.250000  3547.000000
2      AMZN  2020-09-03  3368.000000  3488.409912  3485.000000
3      AMZN  2020-09-04  3294.620117  3381.500000  3318.000000
4      AMZN  2020-09-08  3149.840088  3250.850098  3144.000000

In [41]:
dfkey.head()

symbol                      company                  sector exchange
0   AAPL                   Apple Inc.  Information Technology   NASDAQ
1     VZ  Verizon Communications Inc.  Communication Services     NYSE
2     KO            Coca-Cola Company        Consumer Staples     NYSE
3   AMZN              Amazon.com Inc.  Consumer Discretionary   NASDAQ
4   TSLA                    Tesla Inc  Consumer Discretionary   NASDAQ

In [42]:
# merging in key to the main dataframe
df_merge = pd.merge(df, dfkey, on='symbol', how='left')

In [43]:
df_merge.head()

symbol        date        close         high         open          company  \
0   AMZN  2020-09-01  3499.120117  3513.870117  3489.580078  Amazon.com Inc.   
1   AMZN  2020-09-02  3531.449951  3552.250000  3547.000000  Amazon.com Inc.   
2   AMZN  2020-09-03  3368.000000  3488.409912  3485.000000  Amazon.com Inc.   
3   AMZN  2020-09-04  3294.620117  3381.500000  3318.000000  Amazon.com Inc.   
4   AMZN  2020-09-08  3149.840088  3250.850098  3144.000000  Amazon.com Inc.   

                   sector exchange  
0  Consumer Discretionary   NASDAQ  
1  Consumer Discretionary   NASDAQ  
2  Consumer Discretionary   NASDAQ  
3  Consumer Discretionary   NASDAQ  
4  Consumer Discretionary   NASDAQ

In [44]:
df_merge.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 500 entries, 0 to 499
Data columns (total 8 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   symbol    500 non-null    object 
 1   date      500 non-null    object 
 2   close     500 non-null    float64
 3   high      500 non-null    float64
 4   open      500 non-null    float64
 5   company   500 non-null    object 
 6   sector    500 non-null    object 
 7   exchange  500 non-null    object 
dtypes: float64(3), object(5)
memory usage: 35.2+ KB


In [45]:
len(pd.unique(df_merge['symbol'])) 

5

In [46]:
len(pd.unique(lf_open['symbol'])) 

5

How many unique companies are in the long-format data set? Is that number consistent with the original separated data files?
5 unique companies, and its consistent with the original files

In [47]:
df_merge['date'] = pd.to_datetime( df_merge.date )


In [48]:
df_merge.shape

(500, 8)

How many rows and columns are in your final tidy data set?
8 columns and 500 rows

In [49]:
df_merge.date.value_counts()

2020-12-31    5
2020-09-02    5
2020-10-28    5
2020-12-29    5
2020-10-14    5
             ..
2020-12-02    5
2020-09-17    5
2020-11-18    5
2021-01-19    5
2020-09-15    5
Name: date, Length: 100, dtype: int64

How many rows are associated with each day in the final tidy data set?
15 rows are associated with each day

## Problem 04

In [50]:
df_merge['year'] = df_merge.date.dt.year

df_merge['month'] = df_merge.date.dt.month

df_merge['day'] = df_merge.date.dt.day

df_merge['wday'] = df_merge.date.dt.day_of_week

df_merge['quarter'] = df_merge.date.dt.quarter

In [51]:
df_merge.head()

symbol       date        close         high         open          company  \
0   AMZN 2020-09-01  3499.120117  3513.870117  3489.580078  Amazon.com Inc.   
1   AMZN 2020-09-02  3531.449951  3552.250000  3547.000000  Amazon.com Inc.   
2   AMZN 2020-09-03  3368.000000  3488.409912  3485.000000  Amazon.com Inc.   
3   AMZN 2020-09-04  3294.620117  3381.500000  3318.000000  Amazon.com Inc.   
4   AMZN 2020-09-08  3149.840088  3250.850098  3144.000000  Amazon.com Inc.   

                   sector exchange  year  month  day  wday  quarter  
0  Consumer Discretionary   NASDAQ  2020      9    1     1        3  
1  Consumer Discretionary   NASDAQ  2020      9    2     2        3  
2  Consumer Discretionary   NASDAQ  2020      9    3     3        3  
3  Consumer Discretionary   NASDAQ  2020      9    4     4        3  
4  Consumer Discretionary   NASDAQ  2020      9    8     1        3

In [52]:
df_merge.groupby(['month']).aggregate({'day':'unique'})

day
month                                                   
1      [4, 5, 6, 7, 8, 11, 12, 13, 14, 15, 19, 20, 21...
9      [1, 2, 3, 4, 8, 9, 10, 11, 14, 15, 16, 17, 18,...
10     [1, 2, 5, 6, 7, 8, 9, 12, 13, 14, 15, 16, 19, ...
11     [2, 3, 4, 5, 6, 9, 10, 11, 12, 13, 16, 17, 18,...
12     [1, 2, 3, 4, 7, 8, 9, 10, 11, 14, 15, 16, 17, ...

In [53]:
df_merge.day.nunique()

31

How many unique days are associated with each MONTH in the data set?
31 days

In [54]:
df_merge.head()

symbol       date        close         high         open          company  \
0   AMZN 2020-09-01  3499.120117  3513.870117  3489.580078  Amazon.com Inc.   
1   AMZN 2020-09-02  3531.449951  3552.250000  3547.000000  Amazon.com Inc.   
2   AMZN 2020-09-03  3368.000000  3488.409912  3485.000000  Amazon.com Inc.   
3   AMZN 2020-09-04  3294.620117  3381.500000  3318.000000  Amazon.com Inc.   
4   AMZN 2020-09-08  3149.840088  3250.850098  3144.000000  Amazon.com Inc.   

                   sector exchange  year  month  day  wday  quarter  
0  Consumer Discretionary   NASDAQ  2020      9    1     1        3  
1  Consumer Discretionary   NASDAQ  2020      9    2     2        3  
2  Consumer Discretionary   NASDAQ  2020      9    3     3        3  
3  Consumer Discretionary   NASDAQ  2020      9    4     4        3  
4  Consumer Discretionary   NASDAQ  2020      9    8     1        3

In [55]:
# get difference between open and close 
df_merge['o_c_diff'] =  df_merge['close'] - df_merge['open']

In [56]:
df_merge.head()

symbol       date        close         high         open          company  \
0   AMZN 2020-09-01  3499.120117  3513.870117  3489.580078  Amazon.com Inc.   
1   AMZN 2020-09-02  3531.449951  3552.250000  3547.000000  Amazon.com Inc.   
2   AMZN 2020-09-03  3368.000000  3488.409912  3485.000000  Amazon.com Inc.   
3   AMZN 2020-09-04  3294.620117  3381.500000  3318.000000  Amazon.com Inc.   
4   AMZN 2020-09-08  3149.840088  3250.850098  3144.000000  Amazon.com Inc.   

                   sector exchange  year  month  day  wday  quarter  \
0  Consumer Discretionary   NASDAQ  2020      9    1     1        3   
1  Consumer Discretionary   NASDAQ  2020      9    2     2        3   
2  Consumer Discretionary   NASDAQ  2020      9    3     3        3   
3  Consumer Discretionary   NASDAQ  2020      9    4     4        3   
4  Consumer Discretionary   NASDAQ  2020      9    8     1        3   

     o_c_diff  
0    9.540039  
1  -15.550049  
2 -117.000000  
3  -23.379883  
4    5.840088

In [57]:
df_merge.groupby(['company']).aggregate(o_c_max = ('o_c_diff', 'max'))

o_c_max
company                               
Amazon.com Inc.              95.149902
Coca-Cola Company             1.829998
GameStop Corp.               22.420002
Tesla Inc                    52.700012
Verizon Communications Inc.   1.379997

What is the maximum daily close-to-open difference for each company?
See table above

In [58]:
df_merge.groupby(['company','month']).aggregate(o_c_max = ('o_c_diff', 'max'))

o_c_max
company                     month           
Amazon.com Inc.             1      81.389893
                            9      95.149902
                            10     92.989991
                            11     81.169922
                            12     89.959961
Coca-Cola Company           1       1.050003
                            9       1.829998
                            10      0.750000
                            11      1.410000
                            12      1.069999
GameStop Corp.              1      22.420002
                            9       1.650000
                            10      3.950000
                            11      1.150000
                            12      3.240000
Tesla Inc                   1      38.409973
                            9      38.669983
                            10     11.519989
                            11     38.290009
                            12     52.700012
Verizon Communications Inc. 1       0.949997
                            9       1.379997
                            10      0.599998
                            11      0.819999
                            12      0.689999

What is the maximum daily close-to-open difference for each company in each MONTH?
See table above

In [59]:
df_merge.groupby(['company','month']).aggregate(o_c_max = ('o_c_diff', 'mean'))

o_c_max
company                     month           
Amazon.com Inc.             1      -1.526042
                            9     -19.792841
                            10    -10.758223
                            11     -5.518506
                            12     -2.179066
Coca-Cola Company           1      -0.242000
                            9      -0.036191
                            10     -0.029545
                            11     -0.118500
                            12      0.129545
GameStop Corp.              1       0.972000
                            9       0.064286
                            10      0.031364
                            11      0.024000
                            12      0.051818
Tesla Inc                   1       4.393994
                            9      -0.787619
                            10     -3.280913
                            11      1.339497
                            12      7.314545
Verizon Communications Inc. 1       0.093333
                            9      -0.005714
                            10     -0.081818
                            11      0.024000
                            12     -0.058182

What is the average daily close-to-open difference for each company in each MONTH?
See table above.

In [60]:
df_merge.groupby(['company','day']).aggregate(o_c_max = ('o_c_diff', 'max'))

o_c_max
company                     day           
Amazon.com Inc.             1    31.580078
                            2   -15.550049
                            3    29.879883
                            4    81.169922
                            5    53.359863
...                                    ...
Verizon Communications Inc. 27    0.060002
                            28   -0.180000
                            29    0.599998
                            30    0.390004
                            31    0.689999

[155 rows x 1 columns]

Which DAY had the maximum daily high price for each company in each MONTH?
I think the table above gets at the answer, not quite sure..

In [61]:
df_merge.to_csv('stock_tidy.csv', header=True, index=False)